app/controllers/incomeSources/add/IncomeSourcesAccountingMethodController.scala

```diff
@@ -0,0 +1,245 @@
+    }
+  }
+
+  private def loadIncomeSourceAccountingMethod(isAgent: Boolean, errorHandler: ShowInternalServerError, incomeSourceType: String)
+                                              (implicit user: MtdItUser[_], backUrl: String, postAction: Call, messages: Messages): Future[Result] = {
+    Future.successful(Ok(view(
+      incomeSourcesType = incomeSourceType,
+      form = IncomeSourcesAccountingMethodForm(incomeSourceType),
+      postAction = postAction,
+      isAgent = isAgent,
+      backUrl = backUrl,
+      btaNavPartial = user.btaNavPartial
+    )(user, messages)))
+  }
+
+  def handleRequest(isAgent: Boolean, incomeSourceType: String)
+                   (implicit user: MtdItUser[_], hc: HeaderCarrier, ec: ExecutionContext, messages: Messages): Future[Result] = {
+
+    val incomeSourcesEnabled: Boolean = isEnabled(IncomeSources)
+    val backUrl = incomeSourceType match {
+      case SelfEmployment.key =>
+        if (isAgent)
+            routes.AddBusinessAddressController.showAgent().url
+        else
+            routes.AddBusinessAddressController.show().url
+      case UkProperty.key =>
+        if (isAgent)
+            routes.AddIncomeSourceStartDateCheckController.showUKPropertyAgent.url
+        else
+            routes.AddIncomeSourceStartDateCheckController.showUKProperty.url
+      case ForeignProperty.key =>
+        if (isAgent)
+            routes.AddIncomeSourceStartDateCheckController.showForeignPropertyAgent.url
+        else
+            routes.AddIncomeSourceStartDateCheckController.showForeignProperty.url
+    }
+    val postAction: Call = if (isAgent) controllers.incomeSources.add.routes.IncomeSourcesAccountingMethodController.submitAgent(incomeSourceType) else
+      controllers.incomeSources.add.routes.IncomeSourcesAccountingMethodController.submit(incomeSourceType)
+
+    val errorHandler: ShowInternalServerError = if (isAgent) itvcErrorHandlerAgent else itvcErrorHandler
+
+    if (incomeSourcesEnabled) {
+      if(incomeSourceType == SelfEmployment.key) {
+        handleUserActiveBusinessesCashOrAccruals(isAgent = isAgent, errorHandler = errorHandler, incomeSourceType)(
+          user = user, backUrl = backUrl, postAction = postAction, messages = messages)
+      } else {
+        loadIncomeSourceAccountingMethod(isAgent = isAgent, errorHandler = errorHandler, incomeSourceType)(
+          user = user, backUrl = backUrl, postAction = postAction, messages = messages)
+      }
+    } else {
+      Future.successful(Ok(customNotFoundErrorView()(user, messages)))
+    } recover {
+      case ex: Exception =>
+        Logger("application").error(s"${if (isAgent) "[Agent]"}" +
+          s"Error getting BusinessEndDate page: ${ex.getMessage}")
+        errorHandler.showInternalServerError()
+    }
+  }
+
+  def handleSubmitRequest(isAgent: Boolean, incomeSourceType: String)(implicit user: MtdItUser[_]): Future[Result] = {
+    val (postAction, backUrl, redirect) = incomeSourceType match {
```

Comment:
* you pattern matching over value which then you are using as a parameter, which makes things to complicate to understand;
* Suggested code refactoring:
`
Step A:
  private val actionIndividual =  {
    Map(
      SelfEmployment.key -> (routes.IncomeSourcesAccountingMethodController.submit(SelfEmployment.key),
        routes.AddBusinessAddressController.show().url,
        routes.CheckBusinessDetailsController.show()),
      UkProperty.key -> (routes.IncomeSourcesAccountingMethodController.submit(UkProperty.key),
        routes.AddIncomeSourceStartDateCheckController.showUKProperty.url,
        routes.CheckUKPropertyDetailsController.show()),
      ForeignProperty.key -> (routes.IncomeSourcesAccountingMethodController.submit(ForeignProperty.key),
        routes.AddIncomeSourceStartDateCheckController.showForeignProperty.url,
        routes.ForeignPropertyCheckDetailsController.show())
    )
  }

Step B:
val (postAction, backUrl, redirect) = if (isAgent) {
  actionAgent(incomeSourceType)
} else {
  actionIndividual(incomeSourceType)
}
`